swarm/pss/snapshot_test.go

```diff
@@ -0,0 +1,433 @@
+
+	pof   = pot.DefaultPof(256) // generate messages and index them
+	topic = BytesToTopic([]byte{0x00, 0x00, 0x06, 0x82})
+	mu    sync.Mutex // keeps handlerDonc in sync
+	sim   *simulation.Simulation
+
+	handlerDone   bool // set to true on termination of the simulation run
+	msgsToReceive int  // total count of messages to receive, used for terminating the simulation run
+	maxMessages   int
+	msgCnt        int
+
+	kademlias    map[enode.ID]*network.Kademlia
+	nodeAddrs    map[enode.ID][]byte      // make predictable overlay addresses from the generated random enode ids
+	recipients   map[int][]enode.ID       // for logging output only
+	allowed      map[int][]enode.ID       // allowed recipients
+	expectedMsgs map[enode.ID][]uint64    // message serials we expect respective nodes to receive
+	allowedMsgs  map[enode.ID][]uint64    // message serials we expect respective nodes to receive
+	senders      map[int]enode.ID         // originating nodes of the messages (intention is to choose as far as possible from the receiving neighborhood)
+	handlerC     chan handlerNotification // passes message from pss message handler to simulation driver
+	doneC        chan struct{}            // terminates the handler channel listener
+	errC         chan error               // error to pass to main sim thread
+	msgC         chan handlerNotification // message receipt notification to main sim thread
+	msgs         [][]byte                 // recipient addresses of messages
+)
+
+func resetTestVariables() {
+	handlerDone = false
+	msgsToReceive = 0
+	maxMessages = 0
+	msgCnt = 0
+	msgs = nil
+	sim = nil
+
+	kademlias = make(map[enode.ID]*network.Kademlia)
+	nodeAddrs = make(map[enode.ID][]byte)
+	recipients = make(map[int][]enode.ID)
+	allowed = make(map[int][]enode.ID)
+	expectedMsgs = make(map[enode.ID][]uint64)
+	allowedMsgs = make(map[enode.ID][]uint64)
+	senders = make(map[int]enode.ID)
+	handlerC = make(chan handlerNotification)
+	doneC = make(chan struct{})
+	errC = make(chan error)
+	msgC = make(chan handlerNotification)
+}
+
+func isDone() bool {
+	mu.Lock()
+	defer mu.Unlock()
+	return handlerDone
+}
+
+func setDone() {
+	mu.Lock()
+	defer mu.Unlock()
+	handlerDone = true
+}
+
+func getCmdParams(t *testing.T) (int, int) {
+	args := strings.Split(t.Name(), "/")
+	msgCount, err := strconv.ParseInt(args[1], 10, 16)
+	if err != nil {
+		t.Fatal(err)
+	}
+	nodeCount, err := strconv.ParseInt(args[2], 10, 16)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return int(msgCount), int(nodeCount)
+}
+
+func readSnapshot(t *testing.T, nodeCount int) simulations.Snapshot {
+	f, err := os.Open(fmt.Sprintf("testdata/snapshot_%d.json", nodeCount))
+	if err != nil {
+		t.Fatal(err)
+	}
+	jsonbyte, err := ioutil.ReadAll(f)
+	if err != nil {
+		t.Fatal(err)
+	}
+	var snap simulations.Snapshot
+	err = json.Unmarshal(jsonbyte, &snap)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return snap
+}
+
+func assingTestVariables(sim *simulation.Simulation, msgCount int) {
```

Comment:
I don't understand the logic of when a node is supposed to receive a message and when not. This should be described somewhere (apologies if it's there and I couldn't find or understand it)

Reply:
I agree. The test needs a very clear description on the feature it tests and how it tests it. This is pretty complicated stuff.